doc/man/man1/pattern.1
\"    This documentation is copyright 1994 David Huron.
.TH pattern 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +13
.ti -13
\fBpattern\fR  --  exhaustively locate and count user-defined patterns in a Humdrum input
.in -13
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
.in +9
.ti -9
\fBpattern\fR  [\fB-ciy\fR]  \fB-f \fItemplatefile\fR  [\fB-s \fIregexp\fR]  [\fIinputfile\fR ...]
.in -9
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "pattern"
command is used to locate all occurrences of
a user-defined pattern in some Humdrum input.
The patterns sought may span innumerable data records.
Occurrences of the pattern are identified in the output by line number.
.sp 1
.sp 1
The pattern sought must be defined as a separate \(odtemplate\(cd file.
The template file is identified using the
.B "-f"
command option.
.sp 1
.sp 1
Pattern templates consist of one or more records.
Each record specifies a standard \s-1UNIX\s+1 regular expression --
followed by an optional \(odrecord-count metacharacter.\(cd
The input is scanned from beginning to end, in order to
find passages that match the defined template.
In the simplest case, a match is deemed to take place when successive records
in the input stream match the regular expressions given in each of
the corresponding records of the template.
However, the number of records in the matching input need not be the
same as the number of records in the template.
.sp 1
.sp 1
Consider, first, a simple example where the template consists of
the numbers 1, 2, 3 -- each on a separate line:
.sp 1
.sp 1
.na
.in +2
.nf
1
2
3
.fi
.in -2
.SA
.sp 1
.sp 1
This template will match inputs such as the following:
.sp 1
.sp 1
.in +2
1
.br
112
.br
43.9
.in -2
.sp 1
.sp 1
or even
.in +2
0x.=%&1*
.br
Figure 32
.br
abc(...32...)
.in -2
A more circumspect regular expression template might look like this:
.sp 1
.sp 1
.na
.in +2
.nf
^1$
^2$
^3$
.fi
.in -2
.SA
.sp 1
.sp 1
(The caret (^) and dollar sign ($) are regular expression
.I "anchors"
that indicate the beginning of the record and end of the record
respectively.)
.sp 1
.sp 1
Standard regular expression syntax provides three \(odcounting\(cd
metacharacters that can be used to specify the number of occurrences
of a given pattern on a single line.
The counting metacharacters are +  *  and  ?.
If
.I "p"
is a regular expression pattern, then (\fIp\fR)+ will match
one or more consecutive instances of
.I "p."
Similarly, (\fIp\fR)* will match zero or more consecutive instances of
.I "p,"
whereas (\fIp\fR)? will match zero or one instance of
.I "p."
The use of these metacharacters is illustrated below:
.sp 1
.TS
l l.
X+	matches X, XX, XXX, etc.
X*	matches X, XX, XXX, etc. as well as the null string
X?	matches X or the null string
XX	matches XX
(XX)+	matches XX, XXXX, XXXXXX, etc.
.TE
.sp 1
.ce
.I "Regular expression counting metacharacters"
These metacharacters can be used in conjunction with other regular
expression operators and anchors to specify complex patterns.
See
.B "regexp (6)"
for further details.
.sp 1
.sp 1
In the
.B "pattern"
command, the regular expression counting metacharacters may
.I "also"
be used to specify the
.I "number of successive records"
that match the regular expression.
We refer to this use as \(odrecord-count metacharacters.\(cd
Record-count metacharacters are specified by following the regular expression
with a tab -- followed by either +, *, or ?.
For example, consider the following
.B "pattern"
template:
.in +2
.sp 1
.TS
l l.
X	+
Y	*
Z	?
.TE
.sp 1
.in -2
The intervening tab characters are important here.
They indicate that the metacharacters refer to the number of records
rather than to the number of patterns in a given record.
The first template record (X\fI<tab>\fR+) will match one or more lines
containing the letter X.
The second template record (Y\fI<tab>\fR*) will match zero or more lines
containing the letter Y.
The third template record (Z\fI<tab>\fR?) will match zero or one line
containing the letter Z.
The above template would match an input such as the following:
three successive lines containing the letter X,
followed by eight successive lines containing the letter Y,
followed by a single line containing the letter Z.
Similarly, the above template would also match a one-line input
containing the letter X.
.sp 1
.sp 1
Note that the strings \fI<tab>\fR+ \fI<tab>\fR*,
and \fI<tab>\fR?, are recognized by
.B "pattern"
as record-count metacharacters
.I "only"
if they appear at the end of a regular expression.
.sp 1
.sp 1
The
.B "pattern"
command will identify
.I "all"
possible matching patterns beginning at each point in the input.
Consider, by way of example, the following template file
(named \f(CRtemplate\fR):
.in +2
.sp 1
.TS
l l.
1	+
2	+
3	+
4	+
5	+
.TE
.sp 1
.in -2
The following Humdrum input file is named \f(CRexample1\fR:
.in +2
.sp 1
.TS
l l.
**num	**num
1	1
2	2
3	2
4	3
5	4
5	5
6	6
*-	*-
.TE
.sp 1
.in -2
Given the command:
.sp 1
.sp 1
.in +2
pattern -f template example1
.in -2
.sp 1
.sp 1
the
.B "pattern"
command will produce the following output:
.sp 1
.sp 1
.in +2
4 patterns found from line 2 to line 7 of file example1
.br
1 pattern found from line 2 to line 6 of file example1
.in -2
.sp 1
.sp 1
The patterns are: 1-2-3-4-5, 1-2-2-3-4-5, 1-2-3-3-4-5, 1-2-3-4-4-5 and
1-2-3-4-5-5.
Note that the entire input line is used for matching purposes.
It doesn't matter, for example, whether the number \(od2\(cd is matched
in the left spine or the right spine -- only that the number \(od2\(cd
is present on a given line.
This feature is useful for identifying
.I "Klangfarbenmelodie"
and other \(odthreaded\(cd or \(oddiagonal\(cd patterns that can be
traced between spines.
If the user wishes to avoid such diagonal patterns, individual spines
should be extracted separately before invoking the
.B "pattern"
command.
.in -2
.sp 1
.sp 1
.SH "OPTIONS"
.in +2
The
.B "pattern"
command supports the following options:
.sp 1
.TS
l l.
\fB-c\fR	makes pattern-matching sensitive to comments
\fB-h\fR	displays a help screen summarizing the command
	  syntax
\fB-i\fR	makes pattern-matching sensitive to interpretations
\fB-s \fIregexp\fR	skip (ignore) data records containing the defined
	  regular expression
\fB-y\fR	outputs appropriate `yank' commands in place of
	  regular output
.TE
.sp 1
Options are specified in the command line.
.sp 1
.sp 1
By default, the
.B "pattern"
command is insensitive to the presence or absence of
Humdrum comments and interpretations.
Pattern searches may be made sensitive to occurrences of comments
(defined in the template) by specifying the
.B "-c"
option.
Similarly, pattern searches may be made sensitive to occurrences of
interpretations by specifying the
.B "-i"
option.
.sp 1
.sp 1
Certain types of data records may be ignored in the pattern-search
by invoking the
.B "-s"
(skip) option.
This option must be accompanied by a user-define regular expression.
All input data records matching the regular expression are ignored.
This option is useful, for example, in skipping null data tokens,
barlines, marked embellishment tones, or other types of data.
.sp 1
.sp 1
The
.B "pattern"
command does not directly implement an \(odecho\(cd option
-- such as provided by the related
.B "patt"
command.
With the
.B "-y"
option, however,
.B "pattern"
will produce an output suitable for use with the Humdrum
.B "yank"
command.
This permits the user to extract the appropriate matching passages
from the input.
.in -2
.sp 1
.sp 1
.SH "EXAMPLES"
.in +2
For additional examples pertinent to the
.B "pattern"
command, refer to the \s-1EXAMPLES\s+1 section in the documentation for the
related
.B "patt"
command.
.sp 1
.sp 1
Note that in the above example, the extensive capabilities
for defining complex regular expressions have not been used.
Refer to \fBregexp (6)\fR for further pertinent information.
.in -2
.sp 1
.sp 1
.SH "PORTABILITY"
.in +2
\s-1DOS\s+1 2.0 and up, with the \s-1MKS\s+1 Toolkit.
\s-1OS/2\s+1 with the \s-1MKS\s+1 Toolkit.
\s-1UNIX\s+1 systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fBgrep\fR (\s-1UNIX\s+1), \fBegrep\fR (\s-1UNIX\s+1),
\fBpatt\fR (1), \fBregexp\fR (1),
\fBsimil\fR (1)
.in -2
.sp 1
.sp 1
.SH "WARNINGS"
.in +2
If a comment is present in the template pattern,
failing to specify the
.B "-c"
option will make pattern matching a logically impossibility.
.in -2
.sp 1
.sp 1
.SH "NOTE"
.in +2
The
.B "pattern"
command differs from the related
.B "patt"
command in the following ways:
(1)
.B "patt"
always produces output conforming to the Humdrum syntax whereas
.B "pattern"
never produces Humdrum output.
(2)
.B "patt"
does not support multi-record `wild cards' in the template file,
and so limits the sophistication of the regular expressions.
(3) The
.B "patt"
command provides \(odecho\(cd and \(odtag\(cd options.
.in -2


toolkit/awk/pattern.awk
#####################################################################
##					PATTERN.AWK
# Programmed by: Tim Racinsky         Date: July, 1993
# Copyright (c) 1993 David Huron
#
# Modifications:
#    Date:   	Programmer:		Description:
# May 24/94	Tim Racinsky		Added file name to output
#							Added multiple file capability
# June 7/94	Tim Racinsky		Modified parse_command to work with getopts
#
#
#	This program is called from pattern.ksh and allows the user to search
# for a pattern in the input file specified by a template file.
#
# Options:
#	s regexp - pattern becomes insensitive to data records with the
#		 	 given regular expression 
#	c - makes pattern sensitive to comments
#	i - makes pattern sensitive to interpretations
#	y - outputs 'yank' commands instead of regular output
#
# Main Functions used:
#	Parse_command()	Parse_pattern_array()		Match_record()
#	Check_for_start()	Check_for_finish()			Print_patterns()
#	Find_min()
#
#	VARIABLES:
#  -options:  holds options specified by the user.
#  -reg_exp:  holds regular expression if specified by the user.
#  -smallest_index:  holds smallest possible index of the pattern array that
#   constitutes a matched pattern.
#  -pattern_index:  holds number of patterns in pattern array.
#  -file_name:  holds value of current file name.
#  -first_file:  flat to indicate if this is first file processed or not.
#
#	ARRAYS:
#  -line_repeat:  holds repeat for symbols for each pattern in array pattern.
#  -pattern:  holds the patterns from user defined template.
#  -current_patterns:  holds patterns currently being matched.  This is a 
#   2-dimensional array where the first dimension is a unique pattern index and
#   the second dimension is the line number the pattern started on.  Each
#   array element contains the pattern index of the last pattern matched.
#  -final_array:  2-dimensional array holding the final matched patterns.  The
#   first dimension is the starting line number and the second dimension is the
#   finishing line number.  Each array element holds the number of patterns
#   found between those 2 particular line numbers.
#
BEGIN {
	USAGE="\nUSAGE: pattern -h                 (Help Screen)\n       pattern"\
		 "[-ciy] [-s regexp] -f template [file ...]"
	#
	# Global variables and settings
	#
	FS = OFS = "\t"
	TRUE = 1; FALSE = 0
	#
	# Set standard error to the appropriate 'file'
	#
	#if (ARGV[0] == "nawk") stderr = "/dev/tty"
	#else stderr = "con"
	"echo $CON" | getline stderr
	close("echo $CON")
	options = ""
	reg_exp = ""
	smallest_index = last_index = 0
	pattern_index = 1
	file_name = ""
	first_file = TRUE
	#
	# Read in the user defined template
	#
	while (getline < ARGV[3] > 0)
		{
		last_index++
		pattern[last_index] = $0
		}
	ARGV[3] = ""
	#
	# Determine the user's options and parse the user's template
	#
	parse_command()
	parse_pattern_array()
	}
{
#
# Check to see if a new file has started and if so, reset variables
#
if (FNR == 1)
	{
	if (first_file) first_file = FALSE
	#
	# Print the patterns from the last file
	#
	else
		{
		print_patterns()
		for (i in current_patterns) delete current_patterns[i]
		for (i in final_array) delete final_array[i]
		}
	if (FILENAME == "-") file_name = "STNDIN"
	else file_name = FILENAME
	}
#
# Skip over any lines the user wishes to ignore.  And skip over comments
# and interpretations unless the user specifies otherwise.
#
if (options ~ /s/ && $0 ~ reg_exp) next
else if ($0 ~ /^\!/ && options !~ /c/) next
else if ($0 ~ /^\*/ && options !~ /i/) next
#
# Otherwise, check the current record for the next patterns.
#
else 
	{
	match_record()
	check_for_start()
	check_for_finish()
	}
}
END	{
	#
	# Print the patterns in the appropriate format
	#
	print_patterns()
	}

############################################################################
#				FUNCTIONS USED IN PATTERN.AWK
############################################################################

############################################################################
##				Function Parse_command
# This function inspects the array ARGV in order to determine the correct
# options that the user has specified.
#
function parse_command()
	{
	#
	# Otherwise, set the 'options' variable
	#
	if (ARGV[2] != "null") options = ARGV[2]
	if (ARGV[4] != "null") reg_exp = ARGV[4]
	ARGV[1] = ARGV[2] = ARGV[4] = ""
	}

#########################################################################
##			Function Parse_pattern_array
#	This function loops through the pattern array and checks for any 
# 'repeat line' symbols.  These symbols should be separated from the rest
# of the pattern by a tab and are either a '*', '?', '+', or nothing.
# The array repeat_array is then created depending on these symbols.
#
function parse_pattern_array(     j,arrayb,count,VALID)
	{
	#
	# Loop through each position in the pattern array
	#
	for (j in pattern)
		{
		#
		# Split the pattern array variable to separate the special symbols
		#
		count = split(pattern[j],arrayb)
		#
		# If there is a special symbol, create the appropriate element 
		# in repeat_array
		#
		if (count == 2)
			{
			if (arrayb[2] ~ /\?/) line_repeat[j] = "?"
			else if (arrayb[2] ~ /\+/) line_repeat[j] = "+"
			else if (arrayb[2] ~ /\*/) line_repeat[j] = "*"
			else
				{
				print "pattern: ERROR: Invalid repetition character in "\
					 "expression " pattern[j] "." > stderr
				exit
				}
			}
		else line_repeat[j] = "NULL"
		#
		# If any "\t" characters occur in the pattern, replace them with
		# tab characters
		#
		gsub(/\\t/,"	",arrayb[1])
		pattern[j] = arrayb[1]
		for (j in arrayb) delete arrayb[j]
		}
	#
	# Check that at least one of the patterns specifes that a match
	# must be found (Otherwise, the 'null' line is a match)
	#
	VALID = FALSE
	for (j = last_index; j >= 1; j -= 1)
		{
		if (line_repeat[j] == "+" || line_repeat[j] == "NULL")
			{
			VALID = TRUE
			smallest_index = j
			break
			}
		}
	if (!VALID)
		{
		print "pattern: ERROR: One pattern must specify that at least one\n"\
			 "                match must be found." > stderr
		exit
		}
	}

#############################################################################
##				Function Match_record
#	This function loops through the patterns that are currently being formed
# to see if the current record continues the pattern.
#
function match_record(   i,pattern_number,subscripts,num_of_patterns\
										,current_index,temp_array)
	{
	#
	# Store the indices of the current patterns in a temporary array
	#
	num_of_patterns = 0
	for (i in current_patterns)
		{
		num_of_patterns += 1
		temp_array[num_of_patterns] = i
		}
	#
	# Loop through each current pattern
	#
	for (i = 1; i <= num_of_patterns; i += 1)
		{
		current_index = temp_array[i]
		#
		# Obtain the subscripts of the index
		#
		split(current_index,subscripts,SUBSEP)
		pattern_number = current_patterns[current_index]
		#
		# If the last pattern was a 'multiple' pattern check to see if the
		# current record continues that pattern
		#
		if (line_repeat[pattern_number] == "*" \
							|| line_repeat[pattern_number] == "+")
			{
			if ($0 !~ pattern[pattern_number])
				delete current_patterns[current_index]
			}
		#
		# Otherwise, delete the pattern since it has already been checked
		# in the function check_for_finish()
		#
		else delete current_patterns[current_index]
		#
		# Move on to the next pattern in the pattern array (if it exists)
		#
		pattern_number += 1
		if (pattern_number <= last_index)
			{
			#
			# If the pattern matches, create a new element in the pattern
			# array so that this pattern may be matched
			#
			if ($0 ~ pattern[pattern_number])
				{
				current_patterns[pattern_index,subscripts[2]] \
												= pattern_number
				pattern_index += 1
				}
			#
			# Continue to search for new patterns if the repeat 
			# symbols permit it
			#
			while (line_repeat[pattern_number] == "*" \
								|| line_repeat[pattern_number] == "?")
				{
				pattern_number += 1
				if (pattern_number <= last_index)
					{
					if ($0 ~ pattern[pattern_number])
						{
						current_patterns[pattern_index,subscripts[2]] \
												= pattern_number
						pattern_index += 1
						}
					}
				else break
				}
			}
		}
	}

#############################################################################
##				Function Check_for_start
#	This function checks to see if any patterns start on the current line.
#
function check_for_start(      i)
	{
	#
	# Loop through the array 'pattern'
	#
	for (i = 1; i <= last_index; i += 1)
		{
		#
		# If there is a match, start a new pattern
		#
		if ($0 ~ pattern[i])
			{
			current_patterns[pattern_index,FNR] = i
			pattern_index += 1
			}
		#
		# Continue until the repeat symbols do not permit it any longer
		#
		if (line_repeat[i] != "*" && line_repeat[i] != "?") break
		}
	}

#############################################################################
##				Function Check_for_finish
#	This function checks to see if a pattern can be counted as finished.
#
function check_for_finish(      i,pattern_number,subscripts)
	{
	#
	# Loop through all current patterns
	#
	for (i in current_patterns)
		{
		split(i,subscripts,SUBSEP)
		pattern_number = current_patterns[i]
		#
		# If the pattern's currently matched index qualifies as a 
		# matched pattern, then store it in 'final_array'
		#
		if (pattern_number >= smallest_index)
			{
			if ((subscripts[2],FNR) in final_array)
				final_array[subscripts[2],FNR] += 1
			else
				final_array[subscripts[2],FNR] = 1
			}
		}
	}

####################################################################
##				Function Print_patterns
#	This function is called from the END block and is responsible for
# printing all of the patterns that were found in sequence.
#
function print_patterns(   i,current_index,subscripts)
	{
	current_index = find_min()
	while (current_index != 0)
		{
		split(current_index,subscripts,SUBSEP)
		if (options ~ /y/)
			print "yank -l -r '" subscripts[1] "-" subscripts[2] "'"\
				 " " file_name
		else
			{
			#
			# If only one pattern was found for this particular set of lines,
			# then print that one pattern was found
			#
			if (final_array[current_index] == 1)
				print "1 pattern found from line " subscripts[1] " to "\
				      "line " subscripts[2] " of file " file_name "."
			#
			# Otherwise, print the number of patterns found
			#
			else
				print final_array[current_index] " patterns found from "\
				      "line " subscripts[1] " to line " subscripts[2] \
					 " of file " file_name "."
			}
		delete final_array[current_index]
		#
		# Find the next smallest index.
		#
		current_index = find_min()
		}
	}

###################################################################
##				Function Find_min
#	This function finds the pattern starting at the earliest line number.
#
function find_min(    minimum,i,min_subscripts,subscripts)
	{
	minimum = 0
	for (i in final_array)
		{
		split(i,subscripts,SUBSEP)
		#
		# If minimum has not been set yet, set it to the first element
		#
		if (minimum == 0)
			{
			minimum = i
			min_subscripts = subscripts[1]
			}
		#
		# Otherwise, check to see if the current element (pattern) begins
		# on an earlier line number
		#
		else if (subscripts[1] < min_subscripts)
				{
				minimum = i
				min_subscripts = subscripts[1]
				}
		}
	return minimum
	}
